blockio-sim/src/System/FS/BlockIO/Sim.hs

```diff
@@ -30,11 +34,68 @@ fromHasFS = serialHasBlockIO hSetNoCache hAdvise hAllocate tryLockFile
     hSetNoCache _h _b = pure ()
     hAdvise _ _ _ _ = pure ()
     hAllocate _ _ _ = pure ()
-    -- TODO: implement simulated locks.
-    tryLockFile _ _ = pure Nothing
+
+-- | Lock files are reader\/writer locks.
+--
+-- We implement this using the content of the lock file. The content is a
+-- counter, positive for readers and negaive (specifically -1) for writers.
+-- There can be any number of readers, but only one writer.
+--
+-- Warning: This implementation is not robust under concurrent use (because
+-- operations on files are not atomic) but should be ok for casual use. A
+-- proper implementation would need to be part of the underlying 'HasFs'
+-- implementations.
+--
```

Comment:
We should also maybe warn that the simulation only defines how `tryLockFile`s on the same path interact. A regular `hOpen` on the same path will still work

Reply:
Maybe. The file locking API only interacts with itself, it doesn't block opening files. It's so-called advisory locking, not mandatory locking.